### DISCUSSION TITLE
SHACL with remote repositories?

### DISCUSSION THREAD

**Discussion by uahic** [eclipse-rdf4j/rdf4j#4905] [Q&A] [answered]
Hi there,

is there any way to use SHACL validation if my repository is not running in the same Java VM or more generally, not on the same machine? If my understanding is right, it only works with the SAILS API which in turn is meant to only work locally.

I am using halyard-star (https://github.com/pulquero/Halyard/tree/latest)  (running in a docker container with multiple hbase nodes) which does seem to have a SAILS implementation but if that is only running local on the remote process... meh. 

Thank you in advance
Martin

**Comment by hmottestad** [comment, marked as answer]
This is correct. Our SHACL implementation is designed to work as a protective layer on top of another sail to ensure that no changes to the database can violate the constraints of the shapes. It needs the sail to implement the NotifyingSail interface for this to work. 

Recently we have implemented support for validating data that is stored in an existing database, without support for incremental validation. This still uses the Sail interface, but in theory it could work on the federation sail. I think performance would probably be quite bad though. 

An alternative is to copy all your data into a ShaclSail with a MemoryStore or NativeStore to validate your data. You set up a script to run this process in the night or over the weekend if you have a lot of data.

**Reply by uahic** [reply]
Thanks for the quick and detailed reply! So far we do not have any data in our project as we just started to implementing it. 
Ok this means basically, using for example a HTTPRepository won't give me any access on SHACL error reports remotely if I would add SHACL as an extra layer inside  the halyard rdf4j-server.  I will have a look into the halyard implementation and if I can make our spring-boot server to be the rdf4j server at the same time. Would also solve the problem potentially.

Is SHACL support for SPARQL in the pipe or not feasible due to implementation constraints?

**Reply by hmottestad** [reply]
The validation report is transmitted over http. So you should get hold of that. 

As for SHACL SPARQL we do have support for the basic SPARQL based constraint and SPARQL targets from SHACL Advanced Features Note.
